syntax = "proto3";

package task_service.v1;

option go_package = ".;task_service";

import "google/protobuf/empty.proto";

message TaskListRequest {
    int64 userID = 1;
}

message ProductListItem {
    int64 ID = 1;
    string marketplace = 2;
    int64 SKU = 3;
    string name = 4;
    string photoURL = 5;
    int64 price = 6;
    string brand = 7;
    float rating = 8;
    int64 reviews_count = 9;
    string marketplaceURL = 10;
}

message PhotoShowItem {
    int64 ID = 1;
    string URL = 2;
    bool hasText = 3;
}

message SeoShowItem {
    string keywordID = 1;
    string rawKeyword = 2;
    int64 freq = 3;
    int64 position = 4;
}

message ProductShowItem {
    int64 ID = 1;
    string marketplace = 2;
    int64 SKU = 3;
    string name = 4;
    string description = 5;
    string photoURL = 6;
    repeated PhotoShowItem photos = 7;
    repeated SeoShowItem seo = 8;
    int64 price = 9;
    string brand = 10;
    float rating = 11;
    int64 reviews_count = 12;
    string marketplaceURL = 13;
}

message ProductReviewsResponse {
    repeated string reviews = 1;
}

message TaskListItem {
    int64 ID = 1;
    string url = 2;
    string status = 3;
    optional string errorDetails = 4;
    optional ProductListItem product = 5;
    string createdAt = 6;
    string updatedAt = 7;
}

message ProductTaskShowItem {
    int64 ID = 1;
    string marketplace = 2;
    int64 SKU = 3;
    string name = 4;
    string photoURL = 5;
    string desciption = 6;
    repeated ProductListItem competitors = 7;
    int64 price = 8;
    string brand = 9;
    float rating = 10;
    int64 reviews_count = 11;
    string marketplaceURL = 12;
}

message TaskShowResponse {
    int64 ID = 1;
    string url = 2;
    string status = 3;
    optional string errorDetails = 4;
    optional ProductTaskShowItem product = 5;
    string createdAt = 6;
    string updatedAt = 7;
    string recommendedDescription = 8;
    string reviewsAnalysis = 9;
}

message TaskListResponse {
    int64 userID = 1;
    repeated TaskListItem tasks = 2;
}

message CreateTaskRequest {
    int64 userID = 1;
    string url = 2;
}

message CreateTaskResponse {
    string status = 1;
}

message ShowTaskRequest {
    int64 ID = 1;
}

message ShowProductRequest {
    int64 ID = 1;
}

message ShowReviewsRequest {
    int64 ID = 1;
}

message CheckPhotoData {
    int64 id = 1;
    bool value = 2; 
}

message CheckPhotoRequest {
    bool success = 1;
    string message = 2;
    repeated CheckPhotoData data = 3;
}

message CheckDescriptionData {
    int64 id = 1;
    string value = 2; 
}

message CheckReviewsData {
    int64 id = 1;
    string value = 2; 
}

message UpdateDescriptionRequest {
    bool success = 1;
    string message = 2;
    CheckDescriptionData data = 3;
}

message UpdateReviewsAnalyzeRequest {
    bool success = 1;
    string message = 2;
    CheckReviewsData data = 3;
}

message CheckFontsData {
    int64 id = 1;
    repeated string value = 2; 
}

message UpdateFontsAnalysisRequest {
    bool success = 1;
    string message = 2;
    CheckFontsData data = 3;
}

message CheckQualityData {
    int64 id = 1;
    repeated string value = 2; 
}

message UpdateQualityAnalysisRequest {
    bool success = 1;
    string message = 2;
    CheckQualityData data = 3;
}

message CheckTextOptimizationData {
    int64 id = 1;
    repeated string value = 2; 
}

message UpdateTextOptimizationRequest {
    bool success = 1;
    string message = 2;
    CheckTextOptimizationData data = 3;
}
  
service TaskService {
    rpc List (TaskListRequest) returns (TaskListResponse);
    rpc Create (CreateTaskRequest) returns (CreateTaskResponse);
    rpc CheckPhoto (CheckPhotoRequest) returns (google.protobuf.Empty);
    rpc ShowTask (ShowTaskRequest) returns (TaskShowResponse);
    rpc ShowProduct (ShowProductRequest) returns (ProductShowItem);
    rpc ShowProductReviews (ShowReviewsRequest) returns (ProductReviewsResponse);
    rpc UpdateDescription (UpdateDescriptionRequest) returns (google.protobuf.Empty);
    rpc UpdateReviewsAnalyze (UpdateReviewsAnalyzeRequest) returns (google.protobuf.Empty);
    rpc UpdateFontsAnalysis (UpdateFontsAnalysisRequest) returns (google.protobuf.Empty);
    rpc UpdateQuanlityAnalysis (UpdateQualityAnalysisRequest) returns (google.protobuf.Empty);
    rpc UpdateTextOptimization (UpdateTextOptimizationRequest) returns (google.protobuf.Empty);
}